lib/livebook/utils.ex

```diff
@@ -745,4 +757,68 @@ defmodule Livebook.Utils do
   def ip_to_host({0, 0, 0, 0}), do: "localhost"
   def ip_to_host({127, 0, 0, 1}), do: "localhost"
   def ip_to_host(ip), do: ip |> :inet.ntoa() |> List.to_string()
+
+  @doc """
+  Req plugin adding global Livebook-specific plugs.
+
+  The plugin covers cacerts and HTTP proxy configuration.
+  """
+  @spec req_attach_defaults(Req.Request.t()) :: Req.Request.t()
+  def req_attach_defaults(req) do
+    Req.Request.append_request_steps(req,
+      connect_options: fn request ->
+        uri = URI.parse(request.url)
+        connect_options = mint_connect_options_for_uri(uri)
+        Req.Request.merge_options(request, connect_options: connect_options)
```

Comment:
This is fine as is but if some other code previously set :connect_options, they'd be overridden. I don't have a solution for this in Req, there's just an issue as a reminder: https://github.com/wojtekmach/req/issues/319.

I think ideally it'd be as easy as:

```elixir
put_in(req.options[:connect_options][:proxy], proxy)
```

but it of course assumes the nested structure is already there. I guess what I'm trying to say is it would be nice if there's a language feature that makes above easy or at least if there's a Req feature. WDYT?

Reply:
having a separate function for deep merge is interesting, thanks!